### ISSUE TITLE
Race condition between gather() and AMM

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#7982]
- XREF #4698
- XREF #7993 
- XREF #7995
- XREF #7996

The `who_has` mapping used by `gather` can fall out of sync.
This can happen, for example:

1. when the client is gathering intermediate keys, that are dependencies to other keys, and because of this they are replicated on additional workers while `Scheduler.gather` is running. Afterwards, AMM ReduceReplicas deletes one or more of the excess replicas, which may be the ones in the `who_has` mapping.
2. when a dynamic cluster is shrinking down, which in turn is causing AMM RetireWorkers to create replicas somewhere else and delete the old ones.

When this happens, if a remote worker is missing *any* key, `gather` completely removes the worker from its own internal copy of who_has. It does not refresh who_has from the scheduler.

Once the local copy of `who_has` owned by `gather` runs out of workers for a key, the key is marked as missing, and `gather` fails .

**Comment by crusaderky** [comment]
Correction, in case of `busy` response `gather_from_workers` should outright crash.

**Comment by fjetter** [comment]
@crusaderky thanks for writing this all up but it's a bit hard to parse. The things you are describing appear to be somewhat related because they affect `Client.gather` but their impact and even the affected code are not strictly related.
For instance, I think both "Additional issues" should actually be an additional, dedicated issue each. Can you try to break this down to smaller, easier to digest issues?

> Correction, in case of busy response gather_from_workers should outright crash.

I believe there is an issue for this somewhere already. Unless multiple workloads are running on a cluster, workers should not be busy during Client.gather since they are only replying with busy if there are many concurrent get_data requests.

**Comment by crusaderky** [comment]
Found it: https://github.com/dask/distributed/issues/4698

**Comment by crusaderky** [comment]
I broke down the issue. XREFs in the original post.
